api/api.go

```diff
@@ -48,44 +47,67 @@ func (a *API) ListenAndServe(hostAndPort string) {
 }
 
 // NewAPI instantiates a new REST API
-func NewAPI(config *conf.Configuration, db storage.Connection, mailer mailer.Mailer) *API {
-	return NewAPIWithVersion(context.Background(), config, db, mailer, defaultVersion)
+func NewAPI(globalConfig *conf.GlobalConfiguration, db storage.Connection) *API {
+	return NewAPIWithVersion(context.Background(), globalConfig, db, defaultVersion)
 }
 
 // NewAPIWithVersion creates a new REST API using the specified version
-func NewAPIWithVersion(ctx context.Context, config *conf.Configuration, db storage.Connection, mailer mailer.Mailer, version string) *API {
-	api := &API{config: config, db: db, mailer: mailer, version: version}
+func NewAPIWithVersion(ctx context.Context, globalConfig *conf.GlobalConfiguration, db storage.Connection, version string) *API {
+	api := &API{config: globalConfig, db: db, version: version}
 
 	r := newRouter()
 	r.Use(addRequestID)
 	r.UseBypass(newStructuredLogger(logrus.StandardLogger()))
 	r.Use(recoverer)
 
-	r.Get("/", api.Index)
-	r.Post("/signup", api.Signup)
-	r.Post("/recover", api.Recover)
-	r.Post("/verify", api.Verify)
-	r.Post("/token", api.Token)
-	r.With(api.requireAuthentication).Post("/logout", api.Logout)
-
-	r.Route("/user", func(r *router) {
-		r.Use(api.requireAuthentication)
-		r.Get("/", api.UserGet)
-		r.Put("/", api.UserUpdate)
-	})
-
-	r.Route("/admin", func(r *router) {
-		r.Use(addGetBody)
-		r.Use(api.requireAuthentication)
-		r.Use(api.requireAdmin)
-		r.Get("/users", api.adminUsers)
+	r.Route("/", func(r *router) {
+		if globalConfig.MultiInstanceMode {
+			r.Use(api.loadInstanceConfig)
+		}
 
-		r.Post("/user", api.adminUserCreate)
-		r.Get("/user", api.adminUserGet)
-		r.Put("/user", api.adminUserUpdate)
-		r.Delete("/user", api.adminUserDelete)
+		r.Post("/signup", api.Signup)
+		r.Post("/recover", api.Recover)
+		r.Post("/verify", api.Verify)
+		r.Post("/token", api.Token)
+		r.With(api.requireAuthentication).Post("/logout", api.Logout)
+
+		r.Route("/user", func(r *router) {
+			r.Use(api.requireAuthentication)
+			r.Get("/", api.UserGet)
+			r.Put("/", api.UserUpdate)
+		})
+
+		r.Route("/admin", func(r *router) {
+			r.Use(addGetBody)
+			r.Use(api.requireAuthentication)
+			r.Use(api.requireAdmin)
+
+			r.Get("/users", api.adminUsers)
+
+			r.Post("/user", api.adminUserCreate)
+			r.Get("/user", api.adminUserGet)
+			r.Put("/user", api.adminUserUpdate)
+			r.Delete("/user", api.adminUserDelete)
+		})
 	})
 
+	if globalConfig.MultiInstanceMode {
```

Comment:
I think that you don't have to have to know that you're multi tenant code. If you inject something like a 'configuration manager' or just a middlware that loads the context with the right values the api itself wouldn't know the difference, nor care. 

I think having them separate would give a good separation of concerns, one to manage the configurations one to manage your actual service. I think it is common for someone to take an existing service and just add a management layer on top of it. This would make that part very clear.